src/Workspaces/Remote/Core/Services/AssetService.cs

```diff
@@ -86,23 +88,67 @@ public void UnregisterAssetSource(int serviceId)
 
         public async Task<T> GetAssetAsync<T>(Checksum checksum, CancellationToken cancellationToken)
         {
+            T asset;
+            if (TryGetAsset(checksum, out asset))
+            {
+                return asset;
+            }
+
             using (Logger.LogBlock(FunctionId.AssetService_GetAssetAsync, GetChecksumLogInfo, checksum, cancellationToken))
             {
+                // TODO: what happen if service doesn't come back. timeout?
+                var value = await RequestAssetAsync(checksum, cancellationToken).ConfigureAwait(false);
+
+                _assets.TryAdd(checksum, new Entry(value));
+
+                return (T)value;
+            }
+        }
+
+        public async Task SynchronizeSolutionAssetsAsync(Checksum solutionChecksum, CancellationToken cancellationToken)
+        {
+            using (Logger.LogBlock(FunctionId.AssetService_SynchronizeSolutionAssetsAsync, GetChecksumLogInfo, solutionChecksum, cancellationToken))
+            {
+                var collector = new ChecksumSynchronizer(this);
+                await collector.SynchronizeAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private bool Exists(Checksum checksum)
+        {
+            object unused;
+            return TryGetAsset(checksum, out unused);
```

Comment:
It's probably good that calling 'Exists' ends up updating the timestamp for an item.  After all, that means that somebody needed the item, and so we should try to keep it longer.  However, it's really unclear that that's part of the contract here.  Can you write some comments explaining why we want that.